File: src/aerosurrogate/lifting_line/sections.py
"""Sectional 2D aerodynamic polars for the lifting-line solver.

The lifting-line solver closes its nonlinear system by querying, at every
spanwise station and every iteration, a 2D sectional polar ``Cl(α, Re)`` and
``Cd(α, Re)``. The solver does not care where the polar comes from — only
that it implements the :class:`SectionalAero` protocol.

This module ships three implementations spanning the project's data sources:

* :class:`ThinAirfoilSection` — closed-form ``Cl = 2π(α − α_{L=0})`` from
  ``aerosurrogate.physics``, with an optional parabolic profile-drag model
  ``Cd = Cd0 + k·(Cl − Cl_min)²``. Inviscid in spirit. Used to validate
  the solver against the elliptic-wing analytical identity
  ``CDi = CL² / (π · AR)`` (which holds exactly only when the profile-drag
  contribution is zeroed out).
* :class:`FlatPlatePostStall` — Hoerner ``Cl = sin(2α)`` / ``Cd = 2 sin²(α)``.
  This is *not* a thin-airfoil model in the linear regime, but it has a
  smooth analytical stall at ``α = 45°``, which lets the nonlinear LLT
  solver be exercised through and past stall without any external dependency.
* :class:`NeuralFoilSection` — wraps Peter Sharpe's NeuralFoil (a neural-net
  surrogate for XFOIL) so the LLT solver inherits NeuralFoil's viscous
  panel-method fidelity. This is the production polar; using it here lifts
  the project from "NeuralFoil is a 2D ceiling" to "NeuralFoil sets the
  sectional polar inside a 3D wing solver."
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import Protocol, runtime_checkable

import numpy as np
from numpy.typing import ArrayLike, NDArray

from ..physics import thin_airfoil_cl, thin_airfoil_zero_lift_alpha


@runtime_checkable
class SectionalAero(Protocol):
    """Anything that returns 2D sectional ``Cl`` and ``Cd`` at given ``α``, ``Re``.

    Implementations must be vectorized: passing an array of ``α`` (in degrees)
    and a matching-shape array of ``Re`` must return an array of the same shape.
    """

    def cl(self, alpha_deg: ArrayLike, Re: ArrayLike) -> NDArray[np.float64]:
        ...

    def cd(self, alpha_deg: ArrayLike, Re: ArrayLike) -> NDArray[np.float64]:
        ...


@dataclass
class ThinAirfoilSection:
    """Closed-form thin-airfoil polar ``Cl = 2π(α − α_{L=0})``.

    Lift comes directly from ``aerosurrogate.physics.thin_airfoil_cl`` — the
    same closed-form baseline this repo benchmarks ML models against. Drag is
    modeled as a parabolic polar in ``Cl``::

        Cd = Cd0 + k · (Cl − Cl_min)²

    Setting ``Cd0 = k = 0`` gives a truly inviscid section, which the LLT
    solver uses to recover the analytical identity ``CDi = CL² / (π · AR)``
    on an elliptic wing.

    This class is the natural "physics-only" baseline for the LLT solver —
    parallel to how thin-airfoil theory is the physics-only baseline for the
    2D regression problem in ``physics.py``.

    Stall is not modeled. ``Cl`` grows linearly with ``α`` forever.
    """

    max_camber: float = 0.0
    Cd0: float = 0.0
    k: float = 0.0
    Cl_min_drag: float = 0.0

    @property
    def alpha_L0_deg(self) -> float:
        """Zero-lift angle of attack, degrees."""
        return float(thin_airfoil_zero_lift_alpha(self.max_camber))

    def cl(self, alpha_deg: ArrayLike, Re: ArrayLike | None = None) -> NDArray[np.float64]:  # noqa: ARG002
        return thin_airfoil_cl(alpha_deg, self.max_camber)

    def cd(self, alpha_deg: ArrayLike, Re: ArrayLike | None = None) -> NDArray[np.float64]:
        cl_val = self.cl(alpha_deg, Re)
        return self.Cd0 + self.k * (cl_val - self.Cl_min_drag) ** 2


@dataclass
class FlatPlatePostStall:
    """Hoerner flat-plate polar valid across the full ``[−90°, +90°]`` range::

        Cl(α) = sin(2α)               = 2 sin α cos α
        Cd(α) = Cd0 + 2 sin²(α)

    Reduces to ``Cl ≈ 2α`` (not ``2π α``) at small ``α`` — strictly worse
    than thin-airfoil theory in the linear regime. The purpose of this
    section is *post-stall behavior*: ``Cl`` peaks at ``α = 45°`` and
    decreases beyond, giving the nonlinear LLT solver a sectional polar
    that genuinely stalls so we can demonstrate post-stall wing behavior
    (drop in ``CL``, redistribution of loading toward the unstalled portions).

    Reference: Hoerner, *Fluid-Dynamic Lift*, ch. 4.
    """

    Cd0: float = 0.0

    def cl(self, alpha_deg: ArrayLike, Re: ArrayLike | None = None) -> NDArray[np.float64]:  # noqa: ARG002
        a = np.deg2rad(np.asarray(alpha_deg, dtype=np.float64))
        return np.sin(2.0 * a)

    def cd(self, alpha_deg: ArrayLike, Re: ArrayLike | None = None) -> NDArray[np.float64]:  # noqa: ARG002
        a = np.deg2rad(np.asarray(alpha_deg, dtype=np.float64))
        return self.Cd0 + 2.0 * np.sin(a) ** 2


class NeuralFoilSection:
    """Wrap a NeuralFoil airfoil (XFOIL-grade viscous polar) as a SectionalAero.

    NeuralFoil [#nf]_ is a neural-net surrogate for XFOIL: it returns viscous,
    boundary-layer-resolved 2D sectional aerodynamics in microseconds. Using
    it as the polar inside this lifting-line solver is the whole point of
    the LLT extension — it couples a data-driven 2D viscous model with a
    classical 3D inviscid wing theory, giving a viscous-3D total drag buildup
    cheaper than any panel method.

    Both ``neuralfoil`` and ``aerosandbox`` are *optional* dependencies; the
    imports are lazy so the rest of this package works without them. Install
    with ``pip install -e ".[build]"``.

    Parameters
    ----------
    airfoil_name
        Airfoil identifier resolvable by ``aerosandbox.Airfoil``. Examples:
        ``"naca2412"``, ``"naca0012"``, ``"clarky"``, ``"e387"``.
    model_size
        NeuralFoil model size — one of ``"xxsmall"``, ``"xsmall"``, ``"small"``,
        ``"medium"``, ``"large"``, ``"xlarge"``, ``"xxlarge"``, ``"xxxlarge"``.
        Larger = more accurate, marginally slower. ``"medium"`` is a good default.
    n_crit
        Critical amplification factor (transition criterion). ``9`` is
        standard for clean wind tunnels; ``5`` for noisy / dirty conditions.

    .. [#nf] Sharpe, P. D., "NeuralFoil." MIT-licensed.
       https://github.com/peterdsharpe/NeuralFoil
    """

    def __init__(
        self,
        airfoil_name: str = "naca0012",
        model_size: str = "medium",
        n_crit: float = 9.0,
    ):
        self.airfoil_name = airfoil_name
        self.model_size = model_size
        self.n_crit = float(n_crit)
        # Build the Airfoil once at construction; reuse for every query.
        try:
            import aerosandbox as asb  # noqa: F401
            import neuralfoil  # noqa: F401
        except ImportError as e:  # pragma: no cover
            raise ImportError(
                "NeuralFoilSection requires the optional 'neuralfoil' and "
                "'aerosandbox' packages. Install with: "
                "pip install -e \".[build]\""
            ) from e
        self._airfoil = asb.Airfoil(airfoil_name)

    def _query(self, alpha_deg: ArrayLike, Re: ArrayLike) -> dict:
        import neuralfoil
        a = np.atleast_1d(np.asarray(alpha_deg, dtype=np.float64))
        r = np.broadcast_to(np.asarray(Re, dtype=np.float64), a.shape).copy()
        return neuralfoil.get_aero_from_airfoil(
            airfoil=self._airfoil,
            alpha=a,
            Re=r,
            n_crit=self.n_crit,
            model_size=self.model_size,
        )

    def cl(self, alpha_deg: ArrayLike, Re: ArrayLike) -> NDArray[np.float64]:
        return np.asarray(self._query(alpha_deg, Re)["CL"], dtype=np.float64)

    def cd(self, alpha_deg: ArrayLike, Re: ArrayLike) -> NDArray[np.float64]:
        return np.asarray(self._query(alpha_deg, Re)["CD"], dtype=np.float64)

File: rust/aerosurrogate-core/src/sections.rs
//! Sectional 2D aerodynamic polars for the lifting-line solver.
//!
//! Mirror of Python `aerosurrogate.lifting_line.sections`.
//!
//! The solver closes its nonlinear system by querying, at every spanwise
//! station and every Newton iteration, a 2D sectional polar `Cl(α, Re)` and
//! `Cd(α, Re)`. The solver does not care where the polar comes from — only
//! that it implements the [`SectionalAero`] trait.
//!
//! Implementations in this module:
//!
//! * [`ThinAirfoilSection`] — closed-form `Cl = 2π · (α − α_{L=0})` with an
//!   optional parabolic profile-drag model `Cd = Cd0 + k·(Cl − Cl_min)²`.
//!   Used for the elliptic-wing analytical validation.
//! * [`FlatPlatePostStall`] — Hoerner flat-plate polar `Cl = sin(2α)`,
//!   `Cd = 2sin²(α) + Cd0`. Has an analytical stall at α = 45°, used to
//!   exercise the nonlinear solver post-stall.

use nalgebra::DVector;

/// Anything that returns 2D sectional `Cl` and `Cd` at given α (degrees) and Re.
///
/// Implementations must be vectorized: passing arrays of α and Re must return
/// matching-shape arrays of Cl/Cd. They must also be `Sync` so the solver can
/// hold them behind a `&dyn SectionalAero` reference across thread boundaries
/// (used in benches and the warm-started sweep).
pub trait SectionalAero: Sync {
    /// Sectional lift coefficient at each (α, Re) pair.
    fn cl(&self, alpha_deg: &DVector<f64>, re: &DVector<f64>) -> DVector<f64>;

    /// Sectional drag coefficient at each (α, Re) pair.
    fn cd(&self, alpha_deg: &DVector<f64>, re: &DVector<f64>) -> DVector<f64>;
}

// -------------------------------------------------------------------------
// Thin-airfoil polar
// -------------------------------------------------------------------------

/// Closed-form thin-airfoil polar: `Cl = 2π(α − α_{L=0})`, parabolic drag.
///
/// Lift is exact thin-airfoil theory. Drag is a parabolic polar in `Cl`:
///
/// ```text
/// Cd = Cd0 + k · (Cl − Cl_min)²
/// ```
///
/// Setting `cd0 = k = 0` gives a truly inviscid section, which the LLT
/// solver uses to recover the analytical identity `CDi = CL² / (π · AR)`
/// on an elliptic wing.
///
/// Stall is not modeled. `Cl` grows linearly with α forever.
#[derive(Debug, Default, Clone, Copy)]
pub struct ThinAirfoilSection {
    /// Maximum camber as a fraction of chord. Sets the zero-lift angle.
    pub max_camber: f64,
    /// Zero-lift profile drag coefficient.
    pub cd0: f64,
    /// Induced-of-section / pressure-drag coefficient on the parabolic polar.
    pub k: f64,
    /// Cl at minimum drag (drag bucket center).
    pub cl_min_drag: f64,
}

impl ThinAirfoilSection {
    /// Zero-lift angle of attack (degrees) for the chosen camber.
    ///
    /// For a parabolic camber line `z(x) = 4·m·x·(1−x)` the Glauert integral
    /// collapses to `α_{L=0} = −2·m` in radians.
    pub fn alpha_l0_deg(&self) -> f64 {
        (-2.0 * self.max_camber).to_degrees()
    }
}

impl SectionalAero for ThinAirfoilSection {
    fn cl(&self, alpha_deg: &DVector<f64>, _re: &DVector<f64>) -> DVector<f64> {
        let a_l0 = self.alpha_l0_deg();
        alpha_deg.map(|a| 2.0 * std::f64::consts::PI * (a - a_l0).to_radians())
    }

    fn cd(&self, alpha_deg: &DVector<f64>, re: &DVector<f64>) -> DVector<f64> {
        let cl = self.cl(alpha_deg, re);
        let cd0 = self.cd0;
        let k = self.k;
        let clm = self.cl_min_drag;
        cl.map(|c| cd0 + k * (c - clm).powi(2))
    }
}

// -------------------------------------------------------------------------
// Hoerner flat plate
// -------------------------------------------------------------------------

/// Hoerner flat-plate polar valid across `[−90°, +90°]`:
///
/// ```text
/// Cl(α) = sin(2α)         (≡ 2 sin α cos α)
/// Cd(α) = cd0 + 2 sin²(α)
/// ```
///
/// Reduces to `Cl ≈ 2α` (not `2π α`) at small α — strictly worse than
/// thin-airfoil theory in the linear regime. The purpose of this section
/// is post-stall behavior: `Cl` peaks at α = 45° and decreases beyond,
/// giving the nonlinear LLT solver a polar that genuinely stalls so we
/// can demonstrate post-stall wing behavior.
///
/// Reference: Hoerner, *Fluid-Dynamic Lift*, ch. 4.
#[derive(Debug, Default, Clone, Copy)]
pub struct FlatPlatePostStall {
    /// Zero-lift profile drag added on top of `2 sin²(α)`.
    pub cd0: f64,
}

impl SectionalAero for FlatPlatePostStall {
    fn cl(&self, alpha_deg: &DVector<f64>, _re: &DVector<f64>) -> DVector<f64> {
        alpha_deg.map(|a| (2.0 * a.to_radians()).sin())
    }

    fn cd(&self, alpha_deg: &DVector<f64>, _re: &DVector<f64>) -> DVector<f64> {
        let cd0 = self.cd0;
        alpha_deg.map(|a| cd0 + 2.0 * a.to_radians().sin().powi(2))
    }
}

// -------------------------------------------------------------------------
// Ridge ML surrogate adapter
// -------------------------------------------------------------------------

/// Use the baked-in Poly-2 Ridge surrogate as a `SectionalAero`.
///
/// The surrogate consumes 11 features: 8 geometric (max camber, camber
/// position, max thickness, leading-edge radius, trailing-edge wedge angle,
/// local thickness at 25/50/75% chord) + 3 flow (α, log₁₀ Re, n_crit).
/// Geometric features are fixed at construction so the [`SectionalAero`]
/// interface only varies α (and Re) at runtime.
///
/// This is the headline ML surrogate from the Python `aerosurrogate.models`
/// zoo, baked into the Rust binary as `pub const` arrays via
/// `rust/scripts/export_ridge_weights.py`. No Python interpreter, no sklearn,
/// no allocation in the hot path; one query is ~200 multiply-adds.
#[derive(Debug, Clone, Copy)]
pub struct RidgeSurrogateSection {
    /// Fixed geometric features baked at construction (length 8).
    pub geometry: [f64; 8],
    /// `n_crit` (transition criterion) — varies with conditions but not α.
    pub n_crit: f64,
}

impl RidgeSurrogateSection {
    /// Construct from geometric features in the order expected by the
    /// surrogate: `[max_camber, camber_pos, max_thickness, le_radius,
    /// te_angle_deg, thickness_25, thickness_50, thickness_75]`.
    pub fn from_geometry(geometry: [f64; 8], n_crit: f64) -> Self {
        Self { geometry, n_crit }
    }

    /// Convenience: build the geometry vector for a NACA 4-digit symmetric
    /// airfoil from the 4-digit code. Useful for tests and the demo CLI.
    ///
    /// The geometric features must match how the dataset was built; this
    /// helper applies the same approximation as `aerosurrogate.airfoils`
    /// uses for parametric NACA shapes.
    pub fn naca4(camber_pct: f64, camber_pos_tenth: f64, thickness_pct: f64, n_crit: f64) -> Self {
        let max_camber = camber_pct / 100.0;
        let camber_pos = camber_pos_tenth / 10.0;
        let t = thickness_pct / 100.0;
        // NACA 4-digit thickness distribution: y_t(x) = 5t [0.2969√x - 0.1260x
        // - 0.3516x² + 0.2843x³ - 0.1015x⁴].
        let nact = |x: f64| {
            5.0 * t
                * (0.2969 * x.sqrt() - 0.1260 * x - 0.3516 * x * x
                    + 0.2843 * x.powi(3)
                    - 0.1015 * x.powi(4))
        };
        // LE radius and TE wedge angle approximations consistent with
        // `aerosurrogate.features`.
        let le_radius = 1.1019 * t * t;
        let te_angle_deg = (0.3516_f64).atan().to_degrees(); // ~19.5°, the NACA TE angle.
        let geometry = [
            max_camber,
            camber_pos,
            t,
            le_radius,
            te_angle_deg,
            nact(0.25),
            nact(0.50),
            nact(0.75),
        ];
        Self { geometry, n_crit }
    }

    fn pack_features(&self, alpha_deg: f64, re: f64) -> [f64; 11] {
        let mut f = [0.0_f64; 11];
        f[..8].copy_from_slice(&self.geometry);
        f[8] = alpha_deg;
        f[9] = re.log10();
        f[10] = self.n_crit;
        f
    }
}

impl SectionalAero for RidgeSurrogateSection {
    fn cl(&self, alpha_deg: &DVector<f64>, re: &DVector<f64>) -> DVector<f64> {
        DVector::from_iterator(
            alpha_deg.len(),
            alpha_deg
                .iter()
                .zip(re.iter())
                .map(|(&a, &r)| crate::surrogate::ridge::predict(&self.pack_features(a, r)).cl),
        )
    }

    fn cd(&self, alpha_deg: &DVector<f64>, re: &DVector<f64>) -> DVector<f64> {
        DVector::from_iterator(
            alpha_deg.len(),
            alpha_deg
                .iter()
                .zip(re.iter())
                .map(|(&a, &r)| crate::surrogate::ridge::predict(&self.pack_features(a, r)).cd),
        )
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use approx::assert_relative_eq;
    use std::f64::consts::PI;

    #[test]
    fn thin_airfoil_zero_lift_at_alpha_zero_no_camber() {
        let s = ThinAirfoilSection::default();
        let a = DVector::from_vec(vec![0.0]);
        let r = DVector::from_vec(vec![1e6]);
        assert_relative_eq!(s.cl(&a, &r)[0], 0.0, max_relative = 1e-15);
    }

    #[test]
    fn thin_airfoil_slope_is_two_pi_per_radian() {
        // dCl/dα = 2π/rad. Sample two angles, take secant.
        let s = ThinAirfoilSection::default();
        let r = DVector::from_vec(vec![1e6; 2]);
        let a = DVector::from_vec(vec![0.0, 1.0]); // 0 and 1 degree
        let cl = s.cl(&a, &r);
        let slope_per_deg = cl[1] - cl[0];
        let slope_per_rad = slope_per_deg / 1.0_f64.to_radians();
        assert_relative_eq!(slope_per_rad, 2.0 * PI, max_relative = 1e-12);
    }

    #[test]
    fn thin_airfoil_cambered_has_negative_alpha_l0() {
        let s = ThinAirfoilSection {
            max_camber: 0.02,
            ..Default::default()
        };
        assert!(s.alpha_l0_deg() < 0.0);
    }

    #[test]
    fn flat_plate_peaks_at_45_degrees() {
        let s = FlatPlatePostStall::default();
        let r = DVector::from_vec(vec![1e6; 3]);
        let a = DVector::from_vec(vec![44.0, 45.0, 46.0]);
        let cl = s.cl(&a, &r);
        assert!(cl[1] > cl[0]);
        assert!(cl[1] > cl[2]);
        assert_relative_eq!(cl[1], 1.0, max_relative = 1e-12);
    }

    #[test]
    fn flat_plate_drag_is_positive_everywhere() {
        let s = FlatPlatePostStall { cd0: 0.0 };
        let a: DVector<f64> = DVector::from_iterator(
            181,
            (-90..=90).map(|i| i as f64),
        );
        let r = DVector::from_element(181, 1e6);
        let cd = s.cd(&a, &r);
        for cd_val in cd.iter() {
            assert!(*cd_val >= 0.0);
        }
    }
}
